### ISSUE TITLE
whats the difference between these 2 folders

### ISSUE THREAD

**Issue by waterlike86** [emscripten-core/emsdk#445]
emsdk install 1.39.7 gives 

these 2 folders
emsdk-master\fastcomp\bin\asm2wasm.exe 3mb 
- version from 1.38.+

emsdk-master\upstream\bin\asm2wasm.exe 5.3mb 
- version from 1.39.7

**Comment by waterlike86** [comment]
my mistake i was looking at the wrong version. but what happened between 1.38.+ to 1.39 though. the binary size increased by alot.

**Comment by sbc100** [comment]
That binary comes from binaryen.  How big is the revisions range exactly?  It could just be normal grown in binaryen code size.  If you care about code size you would want to checkout binaryen and build those two revisions and then bisect to find out where the code size increases came from. 

Code size of binaryen tools is not a big concern so we don't monitor it over time.  I guess we assume developers have plenty of disk space.

**Comment by kripken** [comment]
The asm2wasm binary includes all the optimization passes, we keep adding more, to do more optimizations. Could be that.

**Comment by waterlike86** [comment]
strange, i was comparing the difference between 1.38.28 and 1.39.7. the biggest concern is less the size but rather the speed of execution. 
asm2wasm used to take us around 2.5 mins, now takes us 7mins, and wasm-opt and wasm metadce is taking 2 - 3 times longer as well

**Comment by kripken** [comment]
@waterlike86 That does sound odd. I'd bisect this to at least the specific emscripten version, and preferably to the commit - maybe something really did regress here.

**Comment by waterlike86** [comment]
@kripken any luck there?

**Comment by kripken** [comment]
@waterlike86

If you provide me a testcase that shows the 2-3x slower compilation, I can bisect it. If you can't provide me a testcase, please bisect it on your system, or try the system profiler, which might show something.

Testing on some things that I have locally, I don't see a big compile time change (maybe a few % though).

Looking at the asm2wasm binary, I think that's just organic growth as we add more passes and functionality. Here are a bunch of sizes over different versions:
```
1.39.7
-rwxr-xr-x 1 azakai primarygroup 8941016 Feb  2 23:32 fastcomp/bin/asm2wasm

1.39.6
-rwxr-xr-x 1 azakai primarygroup 8751408 Jan 15 06:27 fastcomp/bin/asm2wasm

1.39.5
-rwxr-xr-x 1 azakai primarygroup 8483816 Dec 19 19:09 fastcomp/bin/asm2wasm

1.39.4
-rwxr-xr-x 1 azakai primarygroup 7335048 Dec  2 20:38 fastcomp/bin/asm2wasm

1.39.2
-rwxr-xr-x 1 azakai primarygroup 6926552 Nov  6 09:53 fastcomp/bin/asm2wasm

1.38.45-fastcomp
-rwxr-xr-x 1 azakai primarygroup 6668728 Sep 11  2019 fastcomp/bin/asm2wasm

1.38.33
-rwxr-xr-x 1 azakai primarygroup 5823552 May 23  2019 fastcomp/bin/asm2wasm
```

For example between 1.39.4 and 1.39.5 it looks like we added DWARF support.

Separately, I'd highly recommend upgrading from fastcomp to upstream. Upstream linking should be **massively** faster since it just links, it doesn't compile all the code like fastcomp+asm2wasm do! Any work to optimize asm2wasm would never get it anywhere close to how fast upstream can link.

**Comment by waterlike86** [comment]
@kripken @sbc100  hmm strange. 
I built it myself on windows
binaryen version 91 which i believe its the one used for 1.39.7
this is the size.
23-Apr-20  10:49 PM         3,843,072 asm2wasm.exe
23-Apr-20  10:49 PM         3,979,776 binaryen.dll

i built it using commands
cmake .  -DCMAKE_BUILD_TYPE=Release -DMSVC_USE_LIBCMT=YES
msbuild binaryen.sln -m /property:Configuration=MinSizeRel

this gave us the same build time as 1.38.28 
binary size was
3,133,952 asm2wasm.exe

am i doing something wrong here?

**Comment by kripken** [comment]
I'm not that familiar with building on windows, but perhaps `MinSizeRel` is relevant here? Maybe it makes the build slower but smaller?

**Comment by sbc100** [comment]
Here is how we create the windows binaryies:  https://logs.chromium.org/logs/wasm/buildbucket/cr-buildbucket.appspot.com/8882215311012792720/+/steps/binaryen/0/stdout

We use clang-cl.exe as the compiler so it no surprising that its produces somewhat different results to using MSVC. 

Is there some reason you care about the size of these binaries?    The windows emsdk download is larger that 500Mb so this doesn't seem like a big difference in the grand scale of things.

**Comment by waterlike86** [comment]
> strange, i was comparing the difference between 1.38.28 and 1.39.7. the biggest concern is less the size but rather the speed of execution.
> asm2wasm used to take us around 2.5 mins, now takes us 7mins, and wasm-opt and wasm metadce is taking 2 - 3 times longer as well

yea size was not the matter but the speed of execution. in the end we built binaryen v92 and the speed became closer to that of 1.38.28
asm2wasm used to take us around 2.5 mins, with v92 its taking 3.5mins, much better than v91 which took us 7mins. i think we can live with this timing for now. 

Thanks guys, and sorry for the confusion.

**Comment by sbc100** [comment]
OK, great! 

2.5 to 3.5 minutes still seems like a regression we might want to investigate.  What do you think @tlively @kripken is it time to start recording and test binaryen performance to avoid these kind of cliffs landing by accident?

**Comment by tlively** [comment]
yeah it sounds like this was probably due to the locking issues with `Type` we recently fixed. I think performance tracking would be a very good idea, especially as I continue messing with Binaryen’s fundamental systems. @sbc100 do you have a good idea about how we could do that performance tracking?
